File: app/src/main/java/com/ytempest/wanandroid/activity/main/knowledge/KnowledgeFrag.java
package com.ytempest.wanandroid.activity.main.knowledge;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

import com.ytempest.layoutinjector.annotation.InjectLayout;
import com.ytempest.wanandroid.R;
import com.ytempest.wanandroid.base.fragment.LoaderFrag;
import com.ytempest.wanandroid.base.load.ViewType;
import com.ytempest.wanandroid.http.bean.KnowledgeArchitectureBean;

import java.util.Collections;
import java.util.List;

import butterknife.BindView;

/**
 * @author heqidu
 * @since 2020/6/23
 */
@InjectLayout(R.layout.frag_knowledge)
public class KnowledgeFrag extends LoaderFrag<KnowledgePresenter> implements IKnowledgeContract.View {

    private static final String TAG = KnowledgeFrag.class.getSimpleName();

    @BindView(R.id.group_knowledge_list)
    RecyclerView mRecyclerView;
    private ContentAdapter mAdapter;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mAdapter = new ContentAdapter();
        mRecyclerView.setAdapter(mAdapter);
        loadKnowledgeArchitecture();
    }

    private void loadKnowledgeArchitecture() {
        getLoader().showView(ViewType.LOAD);
        mPresenter.getKnowledgeArchitecture();
    }

    @Override
    protected void onReloadClick() {
        super.onReloadClick();
        loadKnowledgeArchitecture();
    }

    @Override
    public void onKnowledgeArchitectureReceived(List<KnowledgeArchitectureBean> list) {
        getLoader().hideAll();
        Collections.sort(list, (o1, o2) -> Integer.compare(o1.getOrder(), o2.getOrder()));
        mAdapter.display(list);
    }

    @Override
    public void onKnowledgeArchitectureFail(int code) {
        getLoader().showView(ViewType.ERR);
    }
}

File: app/src/main/java/com/ytempest/wanandroid/activity/main/knowledge/KnowledgeFrag.kt
package com.ytempest.wanandroid.activity.main.knowledge

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import com.ytempest.wanandroid.base.createViewModel
import com.ytempest.wanandroid.base.fragment.MVVMFragment
import com.ytempest.wanandroid.base.load.Loader
import com.ytempest.wanandroid.base.load.ViewType
import com.ytempest.wanandroid.base.vm.EntityObserver
import com.ytempest.wanandroid.databinding.FragKnowledgeBinding
import com.ytempest.wanandroid.http.bean.KnowledgeArchitectureBean
import java.util.*

/**
 * @author heqidu
 * @since 21-2-9
 */
class KnowledgeFrag : MVVMFragment<FragKnowledgeBinding>(), IKnowledgeView {

    override val viewModel by lazy { createViewModel<KnowledgeViewModel>() }
    private lateinit var mAdapter: ContentAdapter

    private val loader: Loader by lazy {
        Loader(binding.root as ViewGroup).also {
            it.setReloadCall {
                it.showView(ViewType.LOAD)
                viewModel.loadKnowledgeArchitecture()
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mAdapter = ContentAdapter()
        binding.knowledgeListView.layoutManager = LinearLayoutManager(context)
        binding.knowledgeListView.adapter = mAdapter

        initData()
    }

    private fun initData() {

        viewModel.knowledgeListResult.observe(this, EntityObserver(
            onSuccess = { entity ->
                onKnowledgeArchitectureReceived(entity.data)
            },
            onFail = { entity ->
                onKnowledgeArchitectureFail(entity.code)
            }
        ))

        loader.showView(ViewType.LOAD)
        viewModel.loadKnowledgeArchitecture()
    }

    override fun onKnowledgeArchitectureReceived(list: List<KnowledgeArchitectureBean>) {
        loader.hideAll()
        Collections.sort(list) { p0, p1 -> p0.order.compareTo(p1.order) }
        mAdapter.display(list)
    }

    override fun onKnowledgeArchitectureFail(code: Int) {
        loader.showView(ViewType.ERR);
    }
}